src/strategies.rs

```diff
@@ -0,0 +1,195 @@
+//! This module contains an implementation of the `Hades252`
+//! strategy algorithm specifically designed to work outside of
+//! Rank 1 Constraint Systems (R1CS).
+//!
+//! The inputs of the permutation function have to be explicitly
+//! over the Scalar Field of the curve25519 so working over
+//! `Fp = 2^252 + 27742317777372353535851937790883648493`.
+use crate::mds_matrix::MDS_MATRIX;
+use crate::round_constants::ROUND_CONSTANTS;
+use crate::{PARTIAL_ROUNDS, TOTAL_FULL_ROUNDS};
+use bulletproofs::r1cs::{ConstraintSystem, LinearCombination};
+use curve25519_dalek::scalar::Scalar;
+use std::iter::FromIterator;
+use std::ops::Add;
+
+pub trait StrategyInput = From<Scalar> + Clone + Add;
+
+/// Defines the Hades252 strategy algorithm.
+pub trait Strategy<T>
+where
+    T: StrategyInput,
+    Vec<T>: FromIterator<<T as Add>::Output>,
+{
+    /// Computes `input ^ 5 (mod Fp)`
+    ///
+    /// The modulo depends on the input you use. In our case
+    /// the modulo is done in respect of the `curve25519 scalar field`
+    ///  == `2^252 + 27742317777372353535851937790883648493`.
+    fn quintic_s_box(&mut self, value: &T) -> T;
+
+    /// Multiply the values for MDS matrix.
+    fn mul_matrix(&mut self, values: Vec<T>) -> Vec<T>;
+
+    /// Add round keys to a set of `StrategyInput`.
+    ///
+    /// This round key addition also known as `ARK` is used to
+    /// reach `Confusion and Diffusion` properties for the algorithm.
+    ///
+    /// Basically it allows to destroy any connection between the
+    /// inputs and the outputs of the function.
+    fn add_round_key<'a, I>(&mut self, constants: &mut I, words: Vec<T>) -> Vec<T>
+    where
+        I: Iterator<Item = &'a Scalar>,
+    {
+        words
+            .iter()
+            .map(|word| {
+                let c = constants.next().unwrap();
+                let c = T::from(*c);
+                word.clone() + c
+            })
+            .collect()
+    }
+
+    /// Applies a `Partial Round` also known as a
+    /// `Partial S-Box layer` to a set of inputs.
+    ///
+    /// ### A partial round has 3 steps on every iteration:
+    ///
+    /// - Add round keys to each word. Also known as `ARK`.
+    /// - Apply `quintic S-Box` **just to the first element of
+    /// the words generated from the first step.** This is also known
+    /// as a `Sub Words` operation.
+    /// - Multiplies the output words from the second step by
+    /// the `MDS_MATRIX`.
+    /// This is known as the `Mix Layer`.
+    fn apply_partial_round<'a, I>(&mut self, constants: &mut I, words: Vec<T>) -> Vec<T>
```

Comment:
We reallocate the vec 3 times when we should just mutate the words instead of reallocating

Reply:
Yes, this is just the porting of the original code to the strategies one in order to decouple it.
I feel also this is an improvement that should be done in the follow up PR for the merkle tree integration.

Reply:
Ok!